src/main_solver.jl

```diff
@@ -0,0 +1,121 @@
+function solve!(params::ModelParams)
+    nb = params.nb # Number of blocks
+    x = params.x
+    y = params.y
+    y_L = params.y_L
+    y_U = params.y_U
+
+    ρ = deepcopy(params.opt.step_size_min) # initialize ρ as the minimum allowed step size value
+    μ = 10 # from Boyd's book (make options later)
+    τ = 2
+
+    iter_count = 0
+    params.opt.update_scheme == :JACOBI && (temp_x = zeros(Float64, length(x)))
+    elapsed_time = 0.0
+
+    while !(params.converged || params.tired)
+        iter_count += 1
+        params.opt.update_scheme == :JACOBI && (fill!(temp_x, 0.0))
+        max_iter_time = 0.0
+        params.init_obj_value = 0.0 # reset to zero
+
+        y_slice = @view y[params.model.problem_size.con_counter+1:end]
+
+        for i = 1:nb
+            params.method != DualDecomposition && update_primal!(params.init_blocks[i], x)
+            update_dual!(params.init_blocks[i], y_slice)
+            params.method != DualDecomposition &&
+                params.opt.dynamic_step_size == true &&
+                update_rho!(params.init_blocks[i], ρ)
+
+            optimize_block!(params.init_solver[i], params.results[i])
+
+            if params.opt.update_scheme == :GAUSS_SEIDEL
+                x[params.model.blocks[i].var_idx] = params.results[i].solution
+            elseif params.opt.update_scheme == :JACOBI
+                temp_x[params.model.blocks[i].var_idx] = params.results[i].solution
+            else
+                error("Please choose the update scheme as :JACOBI or :GAUSS_SEIDEL")
+            end
+
+            params.results[i].elapsed_time > max_iter_time &&
+                (max_iter_time = params.results[i].elapsed_time)
+            params.init_obj_value += params.results[i].objective
+            y[params.model.blocks[i].con_idx] = params.results[i].multipliers
+            y_L[params.model.blocks[i].var_idx] = params.results[i].multipliers_L
+            y_U[params.model.blocks[i].var_idx] = params.results[i].multipliers_U
+        end
+        params.opt.update_scheme == :JACOBI && (copyto!(x, temp_x))
+
+        # update res
+        mul!(params.pr_res, params.A, x)
+        axpy!(-1.0, params.b, params.pr_res)
+
+        y[params.model.problem_size.con_counter+1:end] +=
+            params.opt.damping_param * ρ .* params.pr_res
+
+        # update dual_res
+        jac_coord!(params.full_model, x, params.jac[3])
+        coo_prod!(params.jac[2], params.jac[1], params.jac[3], y, params.JTy)
+        grad!(params.full_model, x, params.dl_res)
+        params.dl_res += params.JTy - y_L + y_U
```

Comment:
Maybe use the broadcast operator here, as `dl_res` is a vector: 
```julia
params.dl_res .+= params.JTy .- y_L .+ y_U
```

Reply:
Since I am ensuring that all vectors involved in that operation are of the same dimensions, won't both be equivalent? And won't broadcasting actually be slower?